src/renderers/dom/fiber/ReactDOMFiberComponent.js

```diff
@@ -69,6 +70,66 @@ if (__DEV__) {
     validateInputPropertes(type, props);
     validateUnknownPropertes(type, props);
   };
+
+  var warnForTextDifference = function(serverText: string, clientText: string) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    warning(
+      false,
+      'Text content did not match. Server: "%s" Client: "%s"',
+      serverText,
+      clientText,
+    );
+  };
+
+  var warnForPropDifference = function(
+    propName: string,
+    serverValue: mixed,
+    clientValue: mixed,
+  ) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    warning(
+      false,
+      'Prop `%s` did not match. Server: %s Client: %s',
+      propName,
+      JSON.stringify(serverValue),
+      JSON.stringify(clientValue),
+    );
+  };
+
+  var warnForExtraAttributes = function(attributeNames: Set<string>) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    var names = [];
+    attributeNames.forEach(function(name) {
+      names.push(name);
+    });
```

Comment:
Unfortunately both are pretty slow in V8 at the moment. The fact that `Set.prototype.forEach` is faster in some cases isn't intended, and I wouldn't rely on that. We plan to improve the performance of the collection iterators in the future BTW.

My suggestion here: Don't base decisions on current perf results. Use whatever makes sense for your code and try to improve readability instead.